src/coreclr/src/utilcode/clrhost.cpp

```diff
@@ -12,7 +12,29 @@
 #include "clrnt.h"
 #include "contract.h"
 
-HINSTANCE g_hmodCoreCLR;
+#if HOST_WINDOWS
+extern "C" IMAGE_DOS_HEADER __ImageBase;
+#else
+static void* pImageBase = NULL;
+#endif
+
+void* GetClrModuleBase()
+{
+    LIMITED_METHOD_CONTRACT;
+
+#if HOST_WINDOWS
+    return (void*)&__ImageBase;
+#else // HOST_WINDOWS
+    // PAL_GetSymbolModuleBase defers to dladdr, which is typically a hash lookup through symbols.
+    // It should be fairly fast, however it may take a loader lock, so we will cache the result.
+    if (!pImageBase)
```

Comment:
This is needs to have barrier to prevent the C/C++ compiler from turning this into:
```
void* pRet = pImageBase;
if (pImageBase)
{
   return pRet;
}

pImageBase = (void*)PAL_GetSymbolModuleBase((void*)GetClrModuleBase);
return pImageBase;
```
We had number of subtle race conditions like this over the years...

Reply:
Hmm. It would be strange, but legal transformation. Since `pImageBase` is not volatile (in ISO c++ sense), it can be assumed not changing.

Reply:
Since it is a single location with only one possible transition NULL->value, `LoadWithoutBarrier` should be sufficient.